HW_8/task 8.2.js

```diff
@@ -0,0 +1,269 @@
+
+const enterprises = [
+  {
+    id: 1,
+    name: 'Предприятие 1',
+    departments: [
+      {
+        id: 2,
+        name: 'Отдел тестирования',
+        employees_count: 10,
+      },
+      {
+        id: 3,
+        name: 'Отдел маркетинга',
+        employees_count: 20,
+      },
+      {
+        id: 4,
+        name: 'Администрация',
+        employees_count: 15,
+      },
+    ],
+  },
+  {
+    id: 5,
+    name: 'Предприятие 2',
+    departments: [
+      {
+        id: 6,
+        name: 'Отдел разработки',
+        employees_count: 50,
+      },
+      {
+        id: 7,
+        name: 'Отдел маркетинга',
+        employees_count: 20,
+      },
+      {
+        id: 8,
+        name: 'Отдел охраны труда',
+        employees_count: 5,
+      },
+    ],
+  },
+  {
+    id: 9,
+    name: 'Предприятие 3',
+    departments: [
+      {
+        id: 10,
+        name: 'Отдел аналитики',
+        employees_count: 0,
+      },
+    ],
+  },
+];
+
+// Задания:
+// 1. Вывести все предприятия и их отделы. Рядом указать количество сотрудников.
+// Для предприятия посчитать сумму всех сотрудников во всех отделах.
+
+// **Пример:**
+
+// Предприятие 1 (45 сотрудников)
+// - Отдел тестирования (10 сотрудников)
+// - Отдел маркетинга (20 сотрудников)
+// - Администрация (15 человек)
+// Предприятие 2 (75 сотрудников)
+// - Отдел разработки (50 сотрудников)
+// - Отдел маркетинга (20 сотрудников)
+// - Отдел охраны труда (5 сотрудников)
+// Предприятие 3 (нет сотрудников)
+// - Отдел аналитики (нет сотрудников)
+
+const enterpriseStructure = (enterprises) => {
+    enterprises.map(enterprise => {
+        const totalEmployees = enterprise.departments.reduce((acc, department) => acc + department.employees_count, 0);
+        if (/.*1$/.test(totalEmployees) && !/.*11$/.test(totalEmployees)) {
+            console.log(`${enterprise.name} (${totalEmployees} сотрудник)`);
+        } else {
+        console.log(`${enterprise.name} (${totalEmployees} сотрудников)`);
+        }
+        enterprise.departments.map(department => {
+            if (/.*1$/.test(department.employees_count) && !/.*11$/.test(department.employees_count)) {
+                console.log(`- ${department.name} (${department.employees_count} сотрудник)`);
+            } else {
+            console.log(`- ${department.name} (${department.employees_count} сотрудников)`);
+            }
+        });
+    });
+}
+
+enterpriseStructure(enterprises);
+
+// 2. Написать функцию, которая будет принимать 1 аргумент (id отдела или название отдела и возвращать название предприятия, к которому относится).
+
+// Пример:
+// getEnterpriseName(4) // Предприятие 1
+// getEnterpriseName("Отдел маркетинга") // Предприятие 2
+
+// Функция-помощник для получения предприятия по id отдела или названию отдела:
+const getEnterpriseByDepartment = (idOrName) => enterprises.filter((enterprise) => enterprise.departments.find((department) => department.id === idOrName || department.name === idOrName));
```

Comment:
```suggestion
const getEnterpriseByDepartment = (idOrName) => enterprises.find((enterprise) => enterprise.departments.find((department) => department.id === idOrName || department.name === idOrName));
```
According to functions name, function should return one enterprise, not an array of enterprises

BUT, if you think that there can be more then 1 enterprise with provided department - change function name to getEnterprisesByDepartment

Reply:
Yes, the filter method was chosen because the array has enterprises with the same department name. I'll rename the function, thanks for your suggestion